packages/hardhat-core/src/types/config.ts

```diff
@@ -100,6 +114,17 @@ export interface NetworksConfig {
 
 export type NetworkConfig = HardhatNetworkConfig | HttpNetworkConfig;
 
+export type HardforkHistoryConfig = Map<HardforkName, /* blockNumber */ number>;
```

Comment:
I drafted this up here: https://github.com/nomiclabs/hardhat/pull/2060

Reply:
I'm not sure. I agree that something from `internal` shouldn't be used, but what if we make that enum public? That doesn't sound like a terrible idea.

> Also, maybe we should use string, as it would make this type more future-proof.

Is that so? If we use an enum instead of a string, and then we add support for a new hardfork, the user will have to upgrade hardhat anyway, and so the updated type will be available.

That being said, I don't mind using a string here, provided that:

- When a new hardfork value is added, we get compilation errors in all the right places
- The supported values are validated on runtime.

But I think that will be very close to also having these values typed in the config, and that's a slightly better experience for typescript users, so... maybe just doing that it's better.

Reply:
Oh, I realize now that this is the resolved config, not the user config. I think my comment still applies though, but it would mean also adding it to the user config.

(Anyway, just to be clear: I don't mind using a string, provided we have internal type-safety around the supported hardforks and proper runtime validation, if that lets us merge this sooner)

Reply:
The type-safety around valid (not supported, but valid) hardfork names happens via this: https://github.com/nomiclabs/hardhat/pull/1928/files#diff-cd6765512f8f5e1431a675a3ede5d86714e7665970a45b1f75c97245fdc7cab8R146

The runtime validation happens here: https://github.com/nomiclabs/hardhat/pull/1928/files#diff-f931c37110b4492bfc050d8112ad82e5be4ad8f3ca6bc615ac31cdb19e80453dR2409

Reply:
I'd be ok with a test that leads to the same guarantees that franco is asking for.

Reply:
Addressed with the merging of #2060